From 7be947fd0560d589f0424fb5269d2bda42f3a8cd Mon Sep 17 00:00:00 2001
From: Yurii Rashkovskii <me@yrashk.com>
Date: Tue, 13 Feb 2018 08:36:33 +0700
Subject: [PATCH] Problem: sit-web's /repo/ allows access to the entire disk

For example, http://localhost:8080/repo// (notice the last
slash) allowed to list to the root of the disk and, thus,
lets reads anything.

Solution: make sure the path after `/repo/` resolves to
a directory *within* the actual SIT repository.

Otherwise, return 404
---
 sit-web/src/webapp.rs | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/sit-web/src/webapp.rs b/sit-web/src/webapp.rs
index 88ff4fe..29916b2 100644
--- a/sit-web/src/webapp.rs
+++ b/sit-web/src/webapp.rs
@@ -250,6 +250,9 @@ pub fn start<A: ToSocketAddrs>(addr: A, config: sit_core::cfg::Configuration, re
         // Serve repository content
         if request.url().starts_with("/repo/") {
             let file = repo.path().join(&request.url()[6..]);
+            if file.strip_prefix(repo.path()).is_err() {
+               return Response::empty_404();
+            }
             if file.is_file() {
                 return path_to_response(file)
             } else if file.is_dir() {
-- 
2.15.0

